contracts/HtsSystemContract.sol

```diff
@@ -212,35 +240,106 @@ contract HtsSystemContract is IHederaTokenService, IERC20Events {
+            require(msg.data.length >= 124, "transferFrom: Not enough calldata");
+            address from = address(bytes20(msg.data[40:60]));
+            address to = address(bytes20(msg.data[72:92]));
+            uint256 tokenId = uint256(bytes32(msg.data[92:124]));
+            _transferNFT(from, to, tokenId);
+            return abi.encode(true);
+        }
+        if (selector == IERC721.approve.selector) {
+            require(msg.data.length >= 92, "approve: Not enough calldata");
+            address spender = address(bytes20(msg.data[40:60]));
+            uint256 tokenId = uint256(bytes32(msg.data[60:92]));
+            _approve(spender, tokenId, true);
+            return abi.encode(true);
+        }
+        if (selector == IERC721.setApprovalForAll.selector) {
+            require(msg.data.length >= 92, "setApprovalForAll: Not enough calldata");
+            address operator = address(bytes20(msg.data[40:60]));
+            bool approved = uint256(bytes32(msg.data[60:92])) == 1;
+            _setApprovalForAll(operator, approved);
+            return abi.encode(true);
+        }
+        if (selector == IERC721.getApproved.selector) {
+            require(msg.data.length >= 72, "getApproved: Not enough calldata");
+            uint256 tokenId = uint256(bytes32(msg.data[40:72]));
+            return abi.encode(__getApproved(tokenId));
+        }
+        if (selector == IERC721.isApprovedForAll.selector) {
+            require(msg.data.length >= 92, "isApprovedForAll: Not enough calldata");
+            address owner = address(bytes20(msg.data[40:60]));
+            address operator = address(bytes20(msg.data[72:92]));
+            return abi.encode(__isApprovedForAll(owner, operator));
+        }
+        return __redirectForHRC719(selector);
+    }
+
+    function __redirectForHRC719(bytes4 selector) internal returns (bytes memory) {
+        if (selector == IHRC719.associate.selector) {
             bytes32 slot = _isAssociatedSlot(msg.sender);
             assembly { sstore(slot, true) }
             return abi.encode(true);
-        } else if (selector == IHRC719.dissociate.selector) {
+        }
+        if (selector == IHRC719.dissociate.selector) {
             bytes32 slot = _isAssociatedSlot(msg.sender);
             assembly { sstore(slot, false) }
             return abi.encode(true);
-        } else if (selector == IHRC719.isAssociated.selector) {
+        }
+        if (selector == IHRC719.isAssociated.selector) {
             bytes32 slot = _isAssociatedSlot(msg.sender);
             bool res;
             assembly { res := sload(slot) }
             return abi.encode(res);
-        } else if (selector == this.getTokenInfo.selector) {
+        }
+        revert("redirectForHRC719: not supported");
+    }
+
+    function __redirectForHTS(bytes4 selector) internal virtual returns (bytes memory) {
+        require(msg.sender == HTS_ADDRESS, "redirectForHTS: unauthorized");
```

Comment:
this was already checked when calling this method, right? If the method is `private` then the double-check is not necessary.

Reply:
That is true, but I think it's better to keep methods independent and not introduce coupling by relying on such assumptions, in the long run it's easier for maintaining and less prone to bugs.

Reply:
Nvm, I had to remove `redirectForHTS` and inlined it (without reverting if none of the selectors are matched), because we also want to support ERC-20/ERC-721 method calls when the msg.sender is HTS_ADDRESS

Reply:
"want to support ERC-20/ERC-721 method calls when the msg.sender is HTS_ADDRESS", when does this happen?

Reply:
in mint and burn token:
```solidity
        responseCode = 22; // HederaResponseCodes.SUCCESS
        newTotalSupply = int64(uint64(IERC20(token).totalSupply()));
        serialNumbers = new int64[](0);
        require(newTotalSupply >= 0, "mintToken: invalid total supply");
```

```solidity
        responseCode = 22; // HederaResponseCodes.SUCCESS
        newTotalSupply = int64(uint64(IERC20(token).totalSupply()));
        require(newTotalSupply >= 0, "burnToken: invalid total supply");
```